src/block/block_compact.c

```diff
@@ -65,6 +70,295 @@ __wt_block_compact_get_progress_stats(WT_SESSION_IMPL *session, WT_BM *bm,
+ */
+static inline void
+__block_compact_trim_extent(WT_SESSION_IMPL *session, wt_off_t mask_start, wt_off_t mask_end,
+  wt_off_t *ext_startp, wt_off_t *ext_sizep)
+{
+    WT_UNUSED(session);
+
+    if (mask_end >= 0 && mask_end < mask_start) {
+        *ext_sizep = 0;
+        return;
+    }
+
+    /* Trim from the beginning. */
+    if (*ext_startp < mask_start) {
+        if (*ext_startp + *ext_sizep <= mask_start) {
+            *ext_sizep = 0;
+            return;
+        }
+        *ext_sizep -= mask_start - *ext_startp;
+        *ext_startp = mask_start;
+    }
+
+    /* Trim from the end. */
+    if (mask_end >= 0 && *ext_startp + *ext_sizep > mask_end) {
+        *ext_sizep = mask_end - *ext_startp;
+        if (*ext_sizep <= 0) {
+            *ext_sizep = 0;
+        }
+    }
+}
+
+/*
+ * __block_compact_skip_internal --
+ *     Return if compaction will shrink the file. This function takes a few extra parameters, so
+ *     that it can be useful for both making the actual compaction decisions as well as for
+ *     estimating work ahead of the compaction itself.
+ */
+static void
+__block_compact_skip_internal(WT_SESSION_IMPL *session, WT_BLOCK *block, bool estimate,
+  wt_off_t file_size, wt_off_t start_offset, wt_off_t avail_bytes_before_start_offset, bool *skipp,
+  int *compact_pct_tenths_p)
+{
+    WT_EXT *ext;
+    wt_off_t avail_eighty, avail_ninety, off, size, eighty, ninety;
+
+    /* IMPORTANT: We assume here that block->live_lock is locked. */
+
+    /* Sum the available bytes in the initial 80% and 90% of the file. */
+    avail_eighty = avail_ninety = avail_bytes_before_start_offset;
+    ninety = file_size - file_size / 10;
+    eighty = file_size - ((file_size / 10) * 2);
+
+    WT_EXT_FOREACH_FROM_OFFSET_INCL(ext, &block->live.avail, start_offset)
+    {
+        off = ext->off;
+        size = ext->size;
+        __block_compact_trim_extent(session, start_offset, file_size, &off, &size);
+        if (off < ninety) {
+            avail_ninety += size;
+            if (off < eighty)
+                avail_eighty += size;
+        }
+    }
+
+    /*
+     * Skip files where we can't recover at least 1MB.
+     *
+     * If at least 20% of the total file is available and in the first 80% of the file, we'll try
+     * compaction on the last 20% of the file; else, if at least 10% of the total file is available
+     * and in the first 90% of the file, we'll try compaction on the last 10% of the file.
+     *
+     * We could push this further, but there's diminishing returns, a mostly empty file can be
+     * processed quickly, so more aggressive compaction is less useful.
+     */
+    if (avail_eighty > WT_MEGABYTE && avail_eighty >= ((file_size / 10) * 2)) {
+        *skipp = false;
+        *compact_pct_tenths_p = 2;
+    } else if (avail_ninety > WT_MEGABYTE && avail_ninety >= file_size / 10) {
+        *skipp = false;
+        *compact_pct_tenths_p = 1;
```

Comment:
If we stored 80 and 90 in variables, the two if-statements above could be made more compact.